src/auto-evo/Miche.cs

```diff
@@ -0,0 +1,210 @@
+    {
+        if (Occupant != null)
+        {
+            occupants.Add(Occupant);
+            return occupants;
+        }
+
+        foreach (var child in Children)
+        {
+            occupants.AddRange(child.GetOccupants());
+        }
+
+        return occupants;
+    }
+
+    public IEnumerable<Miche> BackTraversal()
+    {
+        return BackTraversal(new List<Miche>());
+    }
+
+    public IEnumerable<Miche> BackTraversal(List<Miche> currentTraversal)
+    {
+        currentTraversal.Insert(0, this);
+
+        if (Parent == null)
+            return currentTraversal;
+
+        return Parent.BackTraversal(currentTraversal);
+    }
+
+    public void AddChild(Miche newChild)
+    {
+        Children.Add(newChild);
+        newChild.Parent = this;
+    }
+
+    public void AddChildren(IEnumerable<Miche> newChildren)
+    {
+        foreach (var child in newChildren)
+        {
+            AddChild(child);
+        }
+    }
+
+    public bool InsertSpecies(Species species, SimulationCache cache, bool dry = false)
+    {
+        var scoresDictionary = GetOccupants().Distinct().ToDictionary(x => x, _ => 0.0f);
+
+        scoresDictionary[species] = 0.0f;
+
+        return InsertSpecies(species, scoresDictionary, cache, dry);
+    }
+
+    /// <summary>
+    ///   Inserts a species into any spots on the tree where the species is a better fit than any current occupants
+    /// </summary>
+    /// <returns>
+    ///   Returns a bool based on if the species was inserted into a leaf node
+    /// </returns>
+    public bool InsertSpecies(Species species, Dictionary<Species, float> scoresSoFar, SimulationCache cache, bool dry)
+    {
+        var myScore = Pressure.Score(species, cache);
+
+        // Prune branch if species fails any pressures
+        if (myScore <= 0)
+            return false;
+
+        if (IsLeafNode() && Occupant == null)
+        {
+            if (!dry)
+                Occupant = species;
+
+            return true;
+        }
+
+        var newScores = new Dictionary<Species, float>();
+
+        foreach (var currentSpecies in GetOccupants())
+        {
+            newScores[currentSpecies] = scoresSoFar[currentSpecies] +
+                Pressure.WeightedComparedScores(myScore, Pressure.Score(currentSpecies, cache));
+        }
+
+        // We check here to see if scores more than 0, because
+        // scores is relative to the inserted species
+        if (IsLeafNode() && newScores[Occupant!] > 0)
+        {
+            if (!dry)
+                Occupant = species;
+
+            return true;
+        }
+
+        var inserted = false;
+        foreach (var child in Children)
+        {
+            if (child.InsertSpecies(species, newScores, cache, dry))
+            {
+                inserted = true;
+
+                if (dry)
+                    return true;
+            }
+        }
+
+        return inserted;
+    }
+
+    public Miche DeepCopy()
```

Comment:
I don't yet have any suggestion here yet but I dread a bit how much memory allocation this might do. I'll do another go at memory allocation profiling once master is next merged into this.

Reply:
The only comfort I can provide here is that in the latest version of the code that function is only called by the auto-evo exploring tool

Reply:
That's good enough for me, as long as this is not in the main algorithms, I'm fine with having a quick and dirty thing that's just used for GUI stuff outside the main game loop.

Reply:
Sadly I had to bring back the call to this in MutateSpeciesStep. Will need to find a way around that.

Reply:
Once we are down to last few files needing review, I'll try to fully understand the core of the miche alrogithm, and at that point I should hopefully be able to offer design help related to this.